File: src/Asset/Vns.Erp.Core.Asset.CsAssetGUI/Asset/The_TSCD.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using DevExpress.XtraReports.UI;
using DevExpress.XtraPrinting;

using Vns.Erp.Core.Asset.Service.Interface;
using Vns.Erp.Core.Accounting.Service.Interface;
using Vns.Erp.Core.Asset.Domain;
using Vns.Erp.Core.Accounting.Report.Service.Interface;

using Vns.Erp.Core;

namespace Vns.Erp.Core.Asset.CsAssetGUI
{
    public partial class The_TSCD
    {

        #region "Property"
        private IDmTscdService _DmTscdService;
        public IDmTscdService DmTscdService
        {
            get { return _DmTscdService; }
            set { _DmTscdService = value; }
        }

        private IRpTheTscdS23DnService _RpTheTscdS23DnService;
        public IRpTheTscdS23DnService RpTheTscdS23DnService
        {
            get { return _RpTheTscdS23DnService; }
            set { _RpTheTscdS23DnService = value; }
        }

        private ITsTscdPtungService _TsTscdPtungService;
        public ITsTscdPtungService TsTscdPtungService
        {
            get { return _TsTscdPtungService; }
            set { _TsTscdPtungService = value; }
        }
        #endregion

        private string m_TSCD_ID;
        private string LastID = "";
        private int PageNumber = 0;

        private List<DmTscd> lstDmTscd = new List<DmTscd>();
        public The_TSCD(string p_TSCD_ID)
        {
            InitializeComponent();
            BeforePrint += The_TSCD_BeforePrint;
            XrSubreport1.BeforePrint += XrSubreport1_BeforePrint;
            XrSubreport2.BeforePrint += XrSubreport2_BeforePrint;
            m_TSCD_ID = p_TSCD_ID;
        }

        public The_TSCD(string p_TSCD_ID, List<DmTscd> lstTscd)
        {
            InitializeComponent();
            BeforePrint += The_TSCD_BeforePrint;
            XrSubreport1.BeforePrint += XrSubreport1_BeforePrint;
            XrSubreport2.BeforePrint += XrSubreport2_BeforePrint;
            m_TSCD_ID = p_TSCD_ID;
            lstDmTscd = new List<DmTscd>();
            lstDmTscd.AddRange(lstTscd);
        }

        private void XrSubreport2_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
        {
            CT_TSCD prt_ts = (CT_TSCD)((XRSubreport)sender).ReportSource;
            prt_ts.RpTheTscdS23DnService = _RpTheTscdS23DnService;
            prt_ts.DataSource = prt_ts.Bindata(VnsConvert.CGuid(GetCurrentColumnValue("Id")));
        }

        private void XrSubreport1_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
        {
            PTKT rpt_PT = (PTKT)((XRSubreport)sender).ReportSource;
            rpt_PT.TsTscdPtungService = _TsTscdPtungService;
            rpt_PT.DataSource = rpt_PT.GetData(VnsConvert.CGuid(this.GetCurrentColumnValue("Id")));
        }

        private void The_TSCD_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
        {
            DataTable ds_TSCD = DataHelper.ToDataTable(lstDmTscd);
            this.DataSource = ds_TSCD;

            lblDon_VI.Text = Generals.DON_VI.TenDonvi;
            lblDia_Chi.Text = Generals.DON_VI.DiaChi;
        }

        private void xrPage_PrintOnPage(System.Object sender, DevExpress.XtraReports.UI.PrintOnPageEventArgs e)
        {
            //Try
            string CurID = ((XRLabel)sender).Tag.ToString();
            int PageCount = 0;

            foreach (Page MyPage in this.PrintingSystem.Pages)
            {
                BrickEnumerator BrickEnum = MyPage.GetEnumerator();

                while (BrickEnum.MoveNext())
                {
                    VisualBrick Brick = (VisualBrick)BrickEnum.Current;

                    if ((Brick != null) & object.ReferenceEquals(Brick.BrickOwner, (XRLabel)sender))
                    {
                        if (Brick.Value.ToString() == CurID)
                        {
                            PageCount += 1;
                        }
                    }
                }
            }

            if (CurID == LastID)
            {
                PageNumber += 1;
            }
            else
            {
                PageNumber = 1;
            }

            LastID = CurID;
            ((XRLabel)sender).Text = "Page " + PageNumber.ToString() + " of " + PageCount.ToString();
            //Catch ex As Exception
            //End Try
        }
    }
}

File: src/Asset/Vns.Erp.Core.Asset.VbAssetGUI/The_TSCD.vb
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraPrinting

Imports Vns.Erp.Core.Asset.Service.Interface
Imports Vns.Erp.Core.Accounting.Service.Interface
Imports Vns.Erp.Core.Asset.Domain
Imports Vns.Erp.Core.Accounting.Report.Service.Interface

Public Class The_TSCD

#Region "Property"
    Private _DmTscdService As IDmTscdService
    Public Property DmTscdService As IDmTscdService
        Get
            Return _DmTscdService
        End Get
        Set(ByVal value As IDmTscdService)
            _DmTscdService = value
        End Set
    End Property

    Private _RpTheTscdS23DnService As IRpTheTscdS23DnService
    Public Property RpTheTscdS23DnService As IRpTheTscdS23DnService
        Get
            Return _RpTheTscdS23DnService
        End Get
        Set(ByVal value As IRpTheTscdS23DnService)
            _RpTheTscdS23DnService = value
        End Set
    End Property

    Private _TsTscdPtungService As ITsTscdPtungService
    Public Property TsTscdPtungService As ITsTscdPtungService
        Get
            Return _TsTscdPtungService
        End Get
        Set(ByVal value As ITsTscdPtungService)
            _TsTscdPtungService = value
        End Set
    End Property
#End Region

    Private m_TSCD_ID As String
    Private LastID As String = ""
    Private PageNumber As Integer = 0
    Private lstDmTscd As List(Of DmTscd) = New List(Of DmTscd)

    Public Sub New(ByVal p_TSCD_ID As String)
        InitializeComponent()
        m_TSCD_ID = p_TSCD_ID
    End Sub

    Public Sub New(ByVal p_TSCD_ID As String, ByVal lstTscd As List(Of DmTscd))
        InitializeComponent()
        m_TSCD_ID = p_TSCD_ID
        lstDmTscd = New List(Of DmTscd)
        lstDmTscd.AddRange(lstTscd)
    End Sub

    Private Sub XrSubreport2_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles XrSubreport2.BeforePrint
        Dim prt_ts As CT_TSCD = CType((CType(sender, XRSubreport)).ReportSource, CT_TSCD)
        prt_ts.RpTheTscdS23DnService = _RpTheTscdS23DnService
        prt_ts.DataSource = prt_ts.Bindata(GetCurrentColumnValue("Id"))
    End Sub

    Private Sub XrSubreport1_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles XrSubreport1.BeforePrint
        Dim rpt_PT As PTKT = CType((CType(sender, XRSubreport)).ReportSource, PTKT)
        rpt_PT.TsTscdPtungService = _TsTscdPtungService
        rpt_PT.DataSource = rpt_PT.GetData(VnsConvert.CGuid(Me.GetCurrentColumnValue("Id")))
    End Sub

    Private Sub The_TSCD_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles MyBase.BeforePrint
        Dim ds_TSCD As DataTable = DataHelper.ToDataTable(lstDmTscd)
        Me.DataSource = ds_TSCD

        lblDon_VI.Text = Generals.DON_VI.TenDonvi
        lblDia_Chi.Text = Generals.DON_VI.DiaChi
    End Sub

    Private Sub xrPage_PrintOnPage(ByVal sender As System.Object, ByVal e As DevExpress.XtraReports.UI.PrintOnPageEventArgs) Handles xrPage.PrintOnPage
        'Try
        Dim CurID As String = (CType(sender, XRLabel)).Tag.ToString()
        Dim PageCount As Integer = 0

        For Each MyPage As Page In Me.PrintingSystem.Pages
            Dim BrickEnum As BrickEnumerator = MyPage.GetEnumerator()

            Do While BrickEnum.MoveNext()
                Dim Brick As VisualBrick = CType(BrickEnum.Current, VisualBrick)

                If Not (Brick Is Nothing) And Brick.BrickOwner Is CType(sender, XRLabel) Then
                    If Brick.Value.ToString() = CurID Then
                        PageCount += 1
                    End If
                End If
            Loop
        Next MyPage

        If CurID = LastID Then
            PageNumber += 1
        Else
            PageNumber = 1
        End If

        LastID = CurID
        CType(sender, XRLabel).Text = "Page " & PageNumber.ToString() & " of " & PageCount.ToString()
        'Catch ex As Exception
        'End Try
    End Sub
End Class